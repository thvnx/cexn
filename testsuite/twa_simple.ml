(* This file is part of mlexn.

   mlexn is free software: you can redistribute it and/or modify it under the
   terms of the GNU General Public License as published by the Free Software
   Foundation, either version 3 of the License, or (at your option) any later
   version.

   mlexn is distributed in the hope that it will be useful, but WITHOUT ANY
   WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
   FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
   details.

   You should have received a copy of the GNU General Public License along with
   mlexn.  If not, see <http://www.gnu.org/licenses/>. *)

let twa x =
  Printf.printf "add          %s\n" (Twa.to_string (Twa.add (Twa.of_float x) (Twa.of_float x)));
  Printf.printf "sub          %s\n" (Twa.to_string (Twa.sub (Twa.of_float x) (Twa.of_float x)));
  Printf.printf "mul          %s\n" (Twa.to_string (Twa.mul (Twa.of_float x) (Twa.of_float x)));
  Printf.printf "mul_fast     %s\n" (Twa.to_string (Twa.mul_fast (Twa.of_float x) (Twa.of_float x)));
  Printf.printf "mul_dwa      %s\n" (Twa.to_string (Twa.mul_dwa (Twa.of_float x) (Dwa.of_float x)));
  Printf.printf "mul_dwa_fast %s\n" (Twa.to_string (Twa.mul_dwa_fast (Twa.of_float x) (Dwa.of_float x)));
  Printf.printf "reciprocal   %s\n" (Twa.to_string (Twa.reciprocal (Twa.of_float x)));
  Printf.printf "div          %s\n" (Twa.to_string (Twa.div (Twa.of_float x) (Twa.of_float x)));
  Printf.printf "sqrt         %s\n" (Twa.to_string (Twa.sqrt (Twa.of_float x)))

let _ =
  List.iter (fun x ->
      Printf.printf "#TWA tests, x = %h\n" x;
      twa x;
      Printf.printf "\n") [1.; 2.; 4.; 8.; ~-. 1.; ~-. 2.; ~-. 4.; ~-. 8.]
