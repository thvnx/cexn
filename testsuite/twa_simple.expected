#TWA tests, x = 0x1p+0
add          0x1p+1 0x0p+0 0x0p+0
sub          0x0p+0 0x0p+0 0x0p+0
mul          0x1p+0 0x0p+0 0x0p+0
mul_fast     0x1p+0 0x0p+0 0x0p+0
mul_dwa      0x1p+0 0x0p+0 0x0p+0
mul_dwa_fast 0x1p+0 0x0p+0 0x0p+0
reciprocal   0x1p+0 -0x1p-204 0x0p+0
div          0x1p+0 -0x1p-204 0x0p+0
sqrt         0x1p+0 -0x1.2000000000003p-199 0x0p+0

#TWA tests, x = 0x1p+1
add          0x1p+2 0x0p+0 0x0p+0
sub          0x0p+0 0x0p+0 0x0p+0
mul          0x1p+2 0x0p+0 0x0p+0
mul_fast     0x1p+2 0x0p+0 0x0p+0
mul_dwa      0x1p+2 0x0p+0 0x0p+0
mul_dwa_fast 0x1p+2 0x0p+0 0x0p+0
reciprocal   0x1p-1 -0x1p-205 0x0p+0
div          0x1p+0 -0x1p-204 0x0p+0
sqrt         0x1.6a09e667f3bcdp+0 -0x1.bdd3413b26455p-54 -0x1.95abcbb61ea66p-108

#TWA tests, x = 0x1p+2
add          0x1p+3 0x0p+0 0x0p+0
sub          0x0p+0 0x0p+0 0x0p+0
mul          0x1p+4 0x0p+0 0x0p+0
mul_fast     0x1p+4 0x0p+0 0x0p+0
mul_dwa      0x1p+4 0x0p+0 0x0p+0
mul_dwa_fast 0x1p+4 0x0p+0 0x0p+0
reciprocal   0x1p-2 -0x1p-206 0x0p+0
div          0x1p+0 -0x1p-204 0x0p+0
sqrt         0x1p+1 -0x1.2000000000003p-198 0x0p+0

#TWA tests, x = 0x1p+3
add          0x1p+4 0x0p+0 0x0p+0
sub          0x0p+0 0x0p+0 0x0p+0
mul          0x1p+6 0x0p+0 0x0p+0
mul_fast     0x1p+6 0x0p+0 0x0p+0
mul_dwa      0x1p+6 0x0p+0 0x0p+0
mul_dwa_fast 0x1p+6 0x0p+0 0x0p+0
reciprocal   0x1p-3 -0x1p-207 0x0p+0
div          0x1p+0 -0x1p-204 0x0p+0
sqrt         0x1.6a09e667f3bcdp+1 -0x1.bdd3413b26455p-53 -0x1.95abcbb61ea66p-107

#TWA tests, x = -0x1p+0
add          -0x1p+1 0x0p+0 0x0p+0
sub          0x0p+0 0x0p+0 0x0p+0
mul          0x1p+0 0x0p+0 0x0p+0
mul_fast     0x1p+0 0x0p+0 0x0p+0
mul_dwa      0x1p+0 0x0p+0 0x0p+0
mul_dwa_fast 0x1p+0 0x0p+0 0x0p+0
reciprocal   -0x1p+0 0x1p-204 0x0p+0
div          0x1p+0 -0x1p-204 0x0p+0
sqrt         nan -nan -nan

#TWA tests, x = -0x1p+1
add          -0x1p+2 0x0p+0 0x0p+0
sub          0x0p+0 0x0p+0 0x0p+0
mul          0x1p+2 0x0p+0 0x0p+0
mul_fast     0x1p+2 0x0p+0 0x0p+0
mul_dwa      0x1p+2 0x0p+0 0x0p+0
mul_dwa_fast 0x1p+2 0x0p+0 0x0p+0
reciprocal   -0x1p-1 0x1p-205 0x0p+0
div          0x1p+0 -0x1p-204 0x0p+0
sqrt         nan -nan -nan

#TWA tests, x = -0x1p+2
add          -0x1p+3 0x0p+0 0x0p+0
sub          0x0p+0 0x0p+0 0x0p+0
mul          0x1p+4 0x0p+0 0x0p+0
mul_fast     0x1p+4 0x0p+0 0x0p+0
mul_dwa      0x1p+4 0x0p+0 0x0p+0
mul_dwa_fast 0x1p+4 0x0p+0 0x0p+0
reciprocal   -0x1p-2 0x1p-206 0x0p+0
div          0x1p+0 -0x1p-204 0x0p+0
sqrt         nan -nan -nan

#TWA tests, x = -0x1p+3
add          -0x1p+4 0x0p+0 0x0p+0
sub          0x0p+0 0x0p+0 0x0p+0
mul          0x1p+6 0x0p+0 0x0p+0
mul_fast     0x1p+6 0x0p+0 0x0p+0
mul_dwa      0x1p+6 0x0p+0 0x0p+0
mul_dwa_fast 0x1p+6 0x0p+0 0x0p+0
reciprocal   -0x1p-3 0x1p-207 0x0p+0
div          0x1p+0 -0x1p-204 0x0p+0
sqrt         nan -nan -nan

